File: 6kyu/6kyu - Write Number in Expanded Form - Part 2.js
// 6kyu - Write Number in Expanded Form - Part 2

/* This is version 2 of my 'Write Number in Exanded Form' Kata.

You will be given a number and you will need to return it as a string in Expanded Form. For example:

expandedForm(1.24); // should return '1 + 2/10 + 4/100'
expandedForm(7.304); // should return '7 + 3/10 + 4/1000'
expandedForm(0.04); // should return '4/100'         */

// function expandedForm(num) {
//     let arr = String(num).split('.')
//     let int = [...arr[0]].reverse().map((n, i) => n == 0 ? '' : n + '0'.repeat(i)).filter(Boolean).reverse().join(' + ')
//     let dec = [...arr[1]].map((n, i) => n > 0 ? n + '/' + Math.pow(10, i + 1) : '').filter(Boolean).join(' + ')
//     return [int, dec].join(' + ').replace(/^[^0-9]+/, '')
// }

// function expandedForm(num) {
//     let arr = String(num).split('.')
//     let int = [...arr[0]].reverse().map((n, i) => n > 0 ? n + '0'.repeat(i) : '').filter(Boolean).reverse()
//     let dec = [...arr[1]].map((n, i) => n > 0 ? n + '/' + Math.pow(10, i + 1) : '').filter(Boolean)
//     return int.concat(dec).join(' + ')
// }

// function expandedForm(num) {
//     let parts = String(num).split('.')
//     let integer = [...parts[0]].reverse().map((n, i) => n * Math.pow(10, i)).filter(Boolean).reverse()
//     let decimal = [...parts[1]].map((n, i) => n > 0 ? n + '/' + Math.pow(10, i + 1) : '').filter(Boolean)
//     return integer.concat(decimal).join(' + ')
// }


function expandedForm(num) {
    let parts = String(num).split('.')
    let integer = [...parts[0]].reverse().map((n, i) => n * Math.pow(10, i)).reverse()
    let decimal = [...parts[1]].map((n, i) => n > 0 ? n + '/' + Math.pow(10, i + 1) : '')
    return integer.concat(decimal).filter(Boolean).join(' + ')
}


q = expandedForm(1.24) // '1 + 2/10 + 4/100'
q
q = expandedForm(4.28) // '4 + 2/10 + 8/100'
q
q = expandedForm(7.304) // '7 + 3/10 + 4/1000'
q
q = expandedForm(90.304) // '90 + 3/10 + 4/1000'
q
q = expandedForm(91800.304) // '90000 + 1000 + 800 + 3/10 + 4/1000'
q

File: 6kyu/6kyu - Write Number in Expanded Form - Part 2.py
# 6kyu - Write Number in Expanded Form - Part 2

""" You will be given a number and you will need to return it as a string in Expanded Form. """

# def expanded_form(num):
#     integer, decimal = str(num).split('.')
#     decimal = ' + '.join(x+'/1'+'0'*(i+1) for i, x in enumerate(str(decimal)) if int(x))
#     if int(integer):
#         if int(integer) > 9:
#             integer = ' + '.join(x+'0'*(len(str(integer))-i-1) for i, x in enumerate(str(integer)) if int(x))
#             return ' + '.join([integer, decimal])
#         return ' + '.join([integer, decimal])
#     else:
#         return decimal

# def expanded_form(num):
#     parts = str(num).split('.')
#     integer = filter(bool, [int(x) * 10**i for i, x in enumerate(parts[0][::-1])][::-1])
#     decimal = [x+'/' + str(10**(i+1)) for i, x in enumerate(parts[1]) if int(x)]
#     return ' + '.join(list(map(str, integer)) + decimal)


def expanded_form(num):
    parts = str(num).split('.')
    integer = [x + '0' * i for i, x in enumerate(parts[0][::-1]) if int(x)][::-1]
    decimal = [x + '/' + str(10**(i+1)) for i, x in enumerate(parts[1]) if int(x)]
    return ' + '.join(integer + decimal)


q = expanded_form(1.24)  # '1 + 2/10 + 4/100'
q
q = expanded_form(7.304)  # '7 + 3/10 + 4/1000'
q
q = expanded_form(0.04)  # '4/100'
q
q = expanded_form(7909.304)  # '7 + 3/10 + 4/1000'
q
